elastic/kibana
# [Security Solution] Historical rules packages PoC

## Summary

**Related to: https://github.com/elastic/kibana/issues/137420**

I've run a number of tests measuring the performance of the two versions of historical rules packages:

1. A package with a **flat** historical rules structure: a separate saved object for each historical rule version
2. A package with a **composite** historical rules structure: all historical versions of a given rule contained in a single saved object

On a relatively small total number of historical versions (< 10000 total versions, or 10 versions per rule for 1000 rules), the composite structure outperforms the flat one when installing the rules package: 

|  Installation time/package size | 5k versions | 10k versions | 17k versions | 24k versions |
|---|---|---|---|---|
| Flat      | 5.27s / 18M | 7.69 s / 36M | 💥 / 59M | 💥 / 79M | 
| Composite | 2.58s / 16M |  2.97 s / 32M  | 4.11 s / 51M | 4.22 s / 69M |

The difference is visible but not so significant in those numbers. However, things become ugly for the `flat` stricture when we increase the total version number to 15-20k.

### Maximum number of items in a nested field

The first problem that becomes visible is related to the maximum number of items in a nested field. It has already been discussed [here](https://github.com/elastic/kibana/issues/137420#issuecomment-1217779647) and could be easily overcome by adding `enabled: false` to the mappings for the `installed_kibana` field:

https://github.com/elastic/kibana/blob/ab8dd04073a4f9b8bcf26444f35a33d85f98b600/x-pack/plugins/fleet/server/saved_objects/index.ts#L261-L267

### Refresh ran out of slots and forced a refresh

After fixing the above error, the rules package becomes installable, but its installation starts to fail randomly, making Kibana unresponsive for some time. The problem seems to come from the Elasticsearch level. Console logs show dozens of warnings similar to this:

```
block until refresh ran out of slots and forced a refresh: [BulkShardRequest [[.kibana_8.7.0_001][0]] containing [delete {[.kibana_8.7.0][security-rule:d8fc1cca-93ed-43c1-bbb6-c0dd3eff2958:102.0.6]}] blocking until refresh]
```

During that time, all requests to Kibana fail with

```json
{"statusCode":503,"error":"Service Unavailable","message":"connect EADDRNOTAVAIL 127.0.0.1:9200 - Local (0.0.0.0:0)"}
```

But after some time Elasticsearch cluster recovers by itself.

Created a ticket for the Fleet team: https://github.com/elastic/kibana/issues/147695

### All shards failed

Another common error associated with the flat package installation is the following is shards failure. It occurs randomly and is not always easily reproduced:

```json
{"statusCode":503,"error":"Service Unavailable","message":"all shards failed: search_phase_execution_exception: [no_shard_available_action_exception] Reason: null"}
```

After this failure, Elasticsearch doesn't recover by itself, and Kibana responds with:

```json
{"statusCode":503,"error":"Service Unavailable","message":"[No shard available for [get [.kibana_8.7.0][space:default]: routing [null]]: no_shard_available_action_exception: [no_shard_available_action_exception] Reason: No shard available for [get [.kibana_8.7.0][space:default]: routing [null]]]: No shard available for [get [.kibana_8.7.0][space:default]: routing [null]]"}
```

### Response timeout

Sometimes flat package installation just fails with a timeout.

```json
{"statusCode":503,"error":"Service Unavailable","message":"Request timed out"}
```


## Testing instructions

For reference: https://www.elastic.co/guide/en/integrations-developer/current/build-a-new-integration.html

1. Make sure [elastic-package](https://github.com/elastic/elastic-package) CLI and Docker are installed
2. Make sure the original prebuilt rules package is installed (visit any detection engine page to get it installed automatically).
3. If needed, generate more rule versions: `curl -XPOST http://elastic:changeme@localhost:5601/kbn/internal/detection_engine/rules/prebuilt/_install_test_assets -d '{"num_versions_per_rule":10}' -H 'kbn-xsrf: true' -H 'Content-Type: application/json'`
4. Build both flat and composite rules package variations:
   1. `cd fleet-packages/detection-rules-flat && elastic-package build --skip-validation`
   4. `cd fleet-packages/detection-rules-composite && elastic-package build --skip-validation`
5. From the Kibana root, start a copy of the package registry: `elastic-package stack up --services package-registry`
6. Copy the package registry certificate `docker cp <container id>:/etc/ssl/package-registry/ca-cert.pem fleet-packages`
7. Set registry URL in `kibana.dev.yml`: `xpack.fleet.registryUrl: https://localhost:8080`
8. Start Kibana specifying extra certificate `NODE_EXTRA_CA_CERTS=./fleet-packages/ca-cert.pem yarn start`
9. Go to `http://localhost:5601/kbn/app/integrations/browse`. You should find there two detection rules packages: `Prebuilt detection rules (composite)` and `Prebuilt detection rules (flat)`
   <img width="1259" alt="Screenshot 2022-12-08 at 15 32 30" src="https://user-images.githubusercontent.com/1938181/206473237-0fc29959-1d0f-4704-aef2-3eb631288da6.png">
10. To test package installation, use the following requests:
   1. `curl http://elastic:changeme@localhost:5601/kbn/api/fleet/epm/packages/security_rules_flat/8.3.2 -d '{"force":true}' -H 'kbn-xsrf: true' -H 'Content-Type: application/json'`
   3. `curl http://elastic:changeme@localhost:5601/kbn/api/fleet/epm/packages/security_rules_composite/8.3.2 -d '{"force":true}' -H 'kbn-xsrf: true' -H 'Content-Type: application/json'`

## Conclusion

According to this PoC, the composite rule structure looks more stable. However, as outlined in [another PoC](https://github.com/elastic/kibana/pull/144060#pullrequestreview-1224304117), the flat structure provides more benefits when it comes to business logic implementation and overall looks more future-proof. My suggestion would be to fix the current performance issues that are associated with the flat structure and use it as a foundation for the rule customization work.
